File: Algorithms/C++/01984-Minimum Difference Between Highest and Lowest of K Scores.cpp
class Solution {
public:
    int minimumDifference(vector<int>& nums, int k) {
        std::sort(nums.begin(), nums.end());
        int l{0};
        int r = k - 1;
        int min = INT_MAX;

        for(int i{0}; i < nums.size() - k + 1; i++){
            if(r < nums.size()) min = std::min(min, nums[r] - nums[l]);
            l++;
            r++;
        }

        return min;
    }
};

/*
    tc: O(nlogn)
    sc: O(n)

    tc: O(nlogn)
    sc: O(n)
    - sort the array
    - sliding window of size k
    - check min of nums[right] - nums[left]
    
    [1,4,7,100]

*/

File: Algorithms/Python3/01984-Minimum Difference Between Highest and Lowest of K Scores.py
class Solution:
    def minimumDifference(self, nums: List[int], k: int) -> int:
        nums = sorted(nums)
        minDiff = sys.maxsize
    
        for i in range(len(nums) - k + 1):
            minDiff = min(minDiff, nums[k-1+i] - nums[i])
            # window size = k
            # start = nums[i]
            # end = nums[k-1+1]

        return minDiff

'''
    O(nlogn)
    O(1)

[9,4,1,7], k=2
[9,7] = 2
[9,7,4] = 5

sort the array
[1,4,7,9]
 
iterate (size - k + 1) times


pick largest number - 9
pick number at (size - k)th index - 7

return difference of two numbers

'''